packages/guides/src/Compiler/Passes/ReferenceResolverPass.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+declare(strict_types=1);
+
+namespace phpDocumentor\Guides\Compiler\Passes;
+
+use phpDocumentor\Guides\Compiler\CompilerPass;
+use phpDocumentor\Guides\Metas;
+use phpDocumentor\Guides\Nodes\CompoundNode;
+use phpDocumentor\Guides\Nodes\DocumentNode;
+use phpDocumentor\Guides\Nodes\InlineToken\DocReferenceNode;
+use phpDocumentor\Guides\Nodes\InlineToken\ReferenceNode;
+use phpDocumentor\Guides\Nodes\Node;
+use phpDocumentor\Guides\Nodes\SpanNode;
+use phpDocumentor\Guides\UrlGenerator;
+use Psr\Log\LoggerInterface;
+
+use function array_pop;
+use function explode;
+use function implode;
+use function sprintf;
+use function str_contains;
+
+/**
+ * Resolves reference tokens in span nodes
+ *
+ * This follows the reStructuredText rules as outlined in:
+ * https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#implicit-hyperlink-targets
+ */
+class ReferenceResolverPass implements CompilerPass
```

Comment:
I like the movement of this logic to the compiler. That makes a lot of sense. 

What I do see is that you have another case where the NodeTraverser logic doesn't work because you needed the Document. What do you think about adding the `Document` to the parameters you get in `NodeTransformer::enterNode` and `NodeTransformer::leaveNode` I think that would simplify your usecase, as you do not need to have the recursion here. 

Your implementation could be simply

```
class ReferenceResolverNodeTransformer implements NodeTransfomer {

    public function supports(Node $node): bool
    {
        return $node instanceof SpanNode;
    }

     public function enterNode(Node $node, Document $document) {
        return $node;
     }
     
     
     public function leaveNode(Node $node, Document $document) {
              foreach ($child->getTokens() as $token) {
                    if ($token instanceof ReferenceNode) {
                        $this->resolveReference($token, $document);
                    } else {
                        if ($token instanceof DocReferenceNode) {
                            $this->resolveDocReference($token, $document);
                        }
                    }
                }
        
               return $node;
     }
}
```

Reply:
Ok, I see how you mean this. Another Problem however. We first must run all the CollectLinkTargetsTransformers, then the MetasPass and then the reference resolving. Otherwise the link targets are not present yet at the time running the ReferenceNodeTransformer

Reply:
This means we would have to also introduce a priority system for the Transformers...

Reply:
Have a look at my WIP commit